### DISCUSSION TITLE
How Can I derive Temporal Correlation for the Geodetic Receiver Using The Estimated Residuals?

### DISCUSSION THREAD

**Discussion by gnssLa** [groops-devs/groops#89] [Q&A] [answered]
Dear GROOPS Developers,

How Can I derive Temporal Correlation for the Geodetic Receiver Using The Estimated Residuals? For example, Post-Fit residuals?
I know how to calculate the **covarianceEpoch3x3_*** for a given station and the associated output files such as:

- **signalBias_***
- **vtec_***
- **residuals_***
- **position_***

Again, which steps do I need to follow to obtain the **Gaussian noise** for the residuals? 

Thanks for your support and understanding.

Regards,

Bax

**Comment by zhedumi** [comment]
Hi,

Do you intend to compute the temporal correlations of the receiver positions or the temporal correlations of the estimated residuals of the observations?


For the estimated residuals of the observation to derive temporal correlations for a geodetic receiver might require some additional processing steps as has been shown in other research and it is something I am currently also investigating in my research [Dumitraschkewitz 2022](https://graz.pure.elsevier.com/en/activities/empirical-stochastic-modeling-of-observation-noise-in-global-gnss). Some of the methods I apply to derive them are on a delevopment branch in groops and will be available in later releases. 

What you can do though is to use [InstrumentGnssReceiver2TimeSeries](https://groops-devs.github.io/groops/html/InstrumentGnssReceiver2TimeSeries.html), followed by [InstrumentSynchronize](https://groops-devs.github.io/groops/html/InstrumentSynchronize.html)  to split the arcs and finally [Instrument2CovarianceFunctionVCE](https://groops-devs.github.io/groops/html/Instrument2CovarianceFunctionVCE.html).

Best regards,

Patrick

**Comment by gnssLa** [comment]
Hi Patrick,

Thanks for your response.

I am working on both RF and IF level simulators with the target of designing commercial geodetic receivers.  Therefore, the intention is to compute both the temporal correlations of the receiver positions and the temporal correlations of the estimated residuals of the observations. As this will assist in exploring the user limitations of the receivers. 

I went through your research ([Dumitraschkewitz 2022](https://graz.pure.elsevier.com/en/activities/empirical-stochastic-modeling-of-observation-noise-in-global-gnss) ) and I have seen its relevance in what I am trying to validate. I will try to use the 

1. [InstrumentGnssReceiver2TimeSeries](https://groops-devs.github.io/groops/html/InstrumentGnssReceiver2TimeSeries.html),
2. [InstrumentSynchronize](https://groops-devs.github.io/groops/html/InstrumentSynchronize.html)
3. [Instrument2CovarianceFunctionVCE](https://groops-devs.github.io/groops/html/Instrument2CovarianceFunctionVCE.html).

and provide feedback. In the case that I get some issues, I will not hesitate to get back to you.  

In your case, may I know if you applied these steps to the residual output data?

Thanks with regards,

Bax.

**Comment by zhedumi** [comment, marked as answer]
Hi,

In my case yes I also use those steps in my processing with some additional programs which are not yet available in the current release version such as Vondrak filtering, MHM (or Sidereal filtering) filtering and ARMA estimation. The ARMA estimation is a bit different opposed to [Instrument2CovarianceFunctionVCE](https://groops-devs.github.io/groops/html/Instrument2CovarianceFunctionVCE.html) but I think for your purpose this should deliver the same thing since you are interested in the full covariance function and not a model of the covariance function for a stochastic model.
I would recommend going about it in this way actually:

1. [InstrumentGnssReceiver2TimeSeries](https://groops-devs.github.io/groops/html/InstrumentGnssReceiver2TimeSeries.html) create the individual obstype timeseries for each satellite
2. [InstrumentArcCalculate](https://groops-devs.github.io/groops/html/InstrumentArcCalculate.html) remove any outliers from the timeseries
3. [InstrumentSynchronize](https://groops-devs.github.io/groops/html/InstrumentSynchronize.html) split the time series into coherent arcs
4. [InstrumentDetrend](https://groops-devs.github.io/groops/html/InstrumentDetrend.html) remove any trend components
5. [Instrument2CovarianceFunctionVCE](https://groops-devs.github.io/groops/html/Instrument2CovarianceFunctionVCE.html).


For the receiver position I recommend just transfering it into a a local level system with [LocalLevelFrame2StarCamera](https://groops-devs.github.io/groops/html/LocalLevelFrame2StarCamera.html) and [InstrumentRotate](https://groops-devs.github.io/groops/html/InstrumentRotate.html) some [InstrumentDetrend](https://groops-devs.github.io/groops/html/InstrumentDetrend.html) and then do [Instrument2CovarianceFunctionVCE](https://groops-devs.github.io/groops/html/Instrument2CovarianceFunctionVCE.html).



Best regards,

Patrick

**Comment by Waley04** [comment]
Hi Patrick

I was trying a similar issue just like the one presented above for the receiver position. However, the [Instrument2CovarianceFunctionVCE](https://groops-devs.github.io/groops/html/Instrument2CovarianceFunctionVCE.html)  requires **at least 5 arcs for a reliable covariance estimation** and it fails without that information. 

Would you provide a working hint or example on this?

Best regards,

Waley.

**Reply by zhedumi** [reply]
Hi Waley,

Yes in our way of computing the covarianceFunction with VCE we demand at least 5 arcs. To circumvent this for your computation I would suggest splitting the position time series in daily arcs with instrumentSynchronzie or if it is just 1 day to split it in 5 equal arcs for estimation purposes.


Best regards,

Patrick

**Comment by Waley04** [comment]
Hi Patrick,

I went through the source code and I found the threshold of 5 in line#102. 

I have attached a sample script for one station. 

<< [example.zip](https://github.com/groops-devs/groops/files/10029177/example.zip) >>

I was, therefore, wondering if you could show me how to set the parameters using it for the 5 equal arcs. 

Note: 

1.  **receiverSynchronize_2022-05-01.arht === Synchronized File**
2.  **receiverDetrend_2022-05-01.arht** === Detrended File**

Thanks for your support.

Kind regards,

Waley.

**Comment by gnssLa** [comment]
Hi Patrick,

Could you use the example to show the 5 equal arcs as indicated in @Waley04 's issue. 

I am sure that this with help both.

Thanks.

**Comment by zhedumi** [comment]
Hi, 

I am a bit confused the file that you gave is a residual file converted to a timeseries? The provided receiver file is not a position file but a residual file converted to a timeseries. Please note that doing that I would suggest taking into account the individual satellites. In your file within data0 there are all satellites listed leading to the failure of the synchronization ([InstrumentGnssReceiver2TimeSeries](https://groops-devs.github.io/groops/html/InstrumentGnssReceiver2TimeSeries.html)) The instrumentSynchronize splits every epoch which is definitly not something that you want. 
I am going to assume that the detrended values are then also faulty computed since the same timepoints are within the file several times.

So I would suggest first create an individual observation and individual satellite file as suggested in https://github.com/groops-devs/groops/discussions/89#discussioncomment-3328092. In step 1 i meant for each observation and each satellite therefore you should have something like ~30ish residual timeseries file for each satellite. 
To split the data into equal arcs you can just set a maximumArcLength to 150 for example. 

I did a small example with the data you gave me and adapted it to use only one satellite for the time being.
[example.xml.tar.gz](https://github.com/groops-devs/groops/files/10257875/example.xml.tar.gz)

Best regards,

Patrick

**Comment by Waley04** [comment]
@zhedumi, for the sake of curiosity, have you implemented these: 1). Vondrak filtering, 2). MHM (or Sidereal filtering) filtering, and 3). ARMA estimation in the current commit?

**Reply by zhedumi** [reply]
Hi, it is implemented in my development branch but not fit for the main yet since we are discussing refactoring and putting it in a different way.

**Reply by HuLiangliang-skd** [reply]
@zhedumi, with the same curiosity: 1). Vondrak filtering, 2). MHM (or Sidereal filtering) filtering, and 3). ARMA estimation.
Is there open source sharing currently?

**Reply by zhedumi** [reply]
Hi no currently not because I am testing different approaches as well at the moment and before it comes into the code I want to make sure the programms do what I want without bloating up the code unnecessarily.
